src/lib/mina_lib/mina_lib.ml

```diff
@@ -1033,6 +1033,65 @@ let perform_compaction t =
       in
       perform interval_configured
 
+let daemon_start_time = Time_ns.now ()
+
+let check_and_stop_daemon t =
+  let interval_mins = (40 * 60) + (Random.int 16 * 30) in
+  let uptime_mins =
+    Time_ns.(diff (now ()) daemon_start_time |> Span.to_min |> Int.of_float)
+  in
+  let max_catchup_time = Time.Span.of_hr 1. in
+  [%log' info t.config.logger]
+    "Interval minutes $mins Uptime minutes:$uptime_mins "
+    ~metadata:[("mins", `Int interval_mins); ("uptime_mins", `Int uptime_mins)] ;
+  if uptime_mins <= interval_mins then
+    `Check_in
+      (Block_time.Span.to_time_span
+         t.config.precomputed_values.consensus_constants.slot_duration_ms)
+  else
+    match t.next_producer_timing with
+    | None ->
+        `Now
+    | Some timing ->
+        let tm =
+          match timing.timing with
+          | Daemon_rpcs.Types.Status.Next_producer_timing.Check_again tm ->
+              Block_time.to_time tm
+          | Produce tm | Produce_now tm ->
+              Block_time.to_time tm.time
+        in
+        (*Assuming it takes at most 1hr to bootstrap and catchup*)
+        let next_block =
+          Time.add tm
+            (Block_time.Span.to_time_span
+               t.config.precomputed_values.consensus_constants.slot_duration_ms)
+        in
+        let wait_for = Time.(diff next_block (now ())) in
+        if Time.Span.(wait_for > max_catchup_time) then `Now
+        else `Check_in wait_for
+
+let stop_long_running_daemon t =
+  let stop_daemon () =
+    let uptime_mins =
+      Time_ns.(diff (now ()) daemon_start_time |> Span.to_min |> Int.of_float)
+    in
+    [%log' info t.config.logger]
+      "Deamon has been running for $uptime mins. Stopping now..."
+      ~metadata:[("uptime", `Int uptime_mins)] ;
+    Scheduler.yield () >>= (fun () -> exit 0) |> don't_wait_for
+  in
+  let rec go interval =
+    upon (after interval) (fun () ->
+        match check_and_stop_daemon t with
+        | `Now ->
+            stop_daemon ()
+        | `Check_in tm ->
+            go tm )
+  in
+  go
+    (Block_time.Span.to_time_span
+       t.config.precomputed_values.consensus_constants.slot_duration_ms)
```

Comment:
Wouldn't it make more sense to just wait `slot_duration_ms * (40 * 60 + Random.int 16 * 30)` initially before we start the `check_and_stop_daemon` loop?